File: LeetCode/C++/49.cpp
class Solution {
public:
    static bool comp(vector<string> &a, vector<string> &b) {
        return a.size() < b.size();
    }

    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> answer;
        map<string, vector<string>> m;
        
        for(string str : strs) {
            string word = str;
            sort(str.begin(), str.end());
            m[str].push_back(word);
        }
        
        answer.resize(m.size());
        int idx = 0;
        
        for(auto &entry : m) {
            sort(entry.second.begin(), entry.second.end());
            for(auto word : entry.second) {
                answer[idx].push_back(word);
            }
            idx++;
        }
        
        sort(answer.begin(), answer.end(), comp);
        
        return answer;
    }
};

File: LeetCode/JAVA/49.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> answer = new ArrayList<>();
        Map<String, List<String>> m = new HashMap<>();  // MultiMap = key(1) : value(N)

        for(String str : strs) {
            String word = str;
            
            char[] charArr = str.toCharArray();
            Arrays.sort(charArr);
            str = String.valueOf(charArr);
            
            m.computeIfAbsent(str, key -> new ArrayList<String>()).add(word);
        }
        
        for(String key : m.keySet()) {
            answer.add(m.get(key));
        }
        
        return answer;
    }
}